generate.py

```diff
@@ -129,14 +121,125 @@ def generate_message(media, target, conf, result):
     return name, dict(hosts=result, summary=msg)
 
 
+#
+# Sort each target into the respective category
+#
+# targets = All targets with test results
+# categories = All available categories
+def sort_target_into_categories(targets, categories):
+    logger.info("Sorting targets into categories")
+
+    testresults = {}
+
+    for current_category in categories:
+        logger.debug("-> Current category: {:s}".format(current_category))
+
+        testresults[current_category] = {} # Add empty dict
+
+        for current_target in targets:
+            # Check if the current category is listed in the targets category
+            if current_category in targets[current_target]["categories"]:
+                logger.debug("   Target {:s} is in there".format(current_target))
+                testresults[current_category][current_target] = targets[current_target]
+
+    return testresults
+
+#
+# Adds category "Uncategorized" to each target if it has no assigned ones
+#
+def add_uncategorized_category(targets):
+    logger.info("Adding category \"{:s}\" to every target with missing categories".format(KEY_CATEGORY_UNCATEGORIZED))
+
+    for current_target in targets:
+        # Check if element "categories" is in the current target
+        if KEY_CATEGORIES not in targets[current_target]:
+            # Oh no! Add list with single item "Uncategorized"
+            targets[current_target][KEY_CATEGORIES] = ( KEY_CATEGORY_UNCATEGORIZED )
+            logger.debug("-> Added to {:s}".format(current_target))
+
+#
+# Adds a referenceable name key
+# Changes "Telegram Messenger" to "telegrammessenger"
+# Changes "Instagram" to "instagram"
+#
+# This is used in the jinja2 template to build HTML anchors
+#
+def add_referenceable_target_name(targets):
```

Comment:
I would prefer if this function wouldn't mutate the input but instead return a modified dictionary. The interface is then much cleaner and you arguing about (side) effects of code is easier if we stick to a pure-ish function paradigm.

The same applies to `add_uncategorized_category`.

Reply:
Yeah I found out that python passes lists and dicts as reference. So I took advantage of it. Reduces unnecessary memory copies.